File: crawler/crawl_list.php
<?php
/**
 * Phase 1: cào danh sách tất cả bài thuộc 3 category, kèm phân trang.
 * Chỉ insert bài mới (UNIQUE slug) → có thể chạy lại định kỳ để bắt bài mới.
 */

require_once __DIR__ . '/db.php';
require_once __DIR__ . '/http.php';
require_once __DIR__ . '/parser.php';

function crawl_list_run(array $cfg, PDO $pdo, ?array $onlyCategories = null): int
{
    $cats = $onlyCategories ?: $cfg['categories'];
    $newCount = 0;
    foreach ($cats as $cat) {
        $page    = 1;
        $referer = $cfg['base_url'] . $cfg['list_path'];
        $url     = $referer . '?category=' . rawurlencode($cat);

        while ($url !== null) {
            db_log($pdo, 'list', "Fetch $url");
            // Listing page hay đổi (có thêm bài mới) nên không cache:
            $html   = http_fetch($cfg, $url, $referer, false);
            $parsed = parse_listing($html);
            $added  = 0;
            foreach ($parsed['lessons'] as $l) {
                $l['category_slug'] = $cat;
                if (db_upsert_lesson_listing($pdo, $l)) {
                    $added++;
                    $newCount++;
                }
            }
            db_log($pdo, 'list', "  category=$cat page=$page found=" . count($parsed['lessons']) . " new=$added");

            if (!$parsed['next_url'] || $parsed['next_url'] === $url) {
                break;
            }
            $referer = $url;
            $url     = $parsed['next_url'];
            $page++;
            // Hard cap để tránh loop bất tận do site đổi pagination:
            if ($page > 50) {
                db_log($pdo, 'list', "  WARN: dừng vì >50 trang, kiểm tra thủ công");
                break;
            }
        }
    }
    db_log($pdo, 'list', "Tổng số bài mới thêm: $newCount");
    return $newCount;
}

if (PHP_SAPI === 'cli' && realpath($_SERVER['SCRIPT_FILENAME']) === __FILE__) {
    $cfg = require __DIR__ . '/config.php';
    $pdo = db_open($cfg);
    $only = null;
    if (isset($argv[1])) {
        $only = explode(',', $argv[1]);
    }
    crawl_list_run($cfg, $pdo, $only);
}

File: crawler_py/crawl_list.py
"""Phase 1: cào danh sách bài (đơn luồng)."""
from __future__ import annotations

# Cho phép chạy trực tiếp `python crawl_list.py` lẫn `python -m crawler_py.crawl_list`.
if __name__ == "__main__" and __package__ in (None, ""):
    import os, sys
    sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
    __package__ = "crawler_py"

from urllib.parse import quote

from . import config, db, httpclient, parser


def run(category_filter: list[str] | None = None) -> int:
    conn = db.open_conn()
    session = httpclient.make_session()

    cats = category_filter or config.CATEGORIES
    new_total = 0
    for cat in cats:
        page = 1
        referer = config.BASE_URL + config.LIST_PATH
        url = f"{referer}?category={quote(cat)}"

        while url:
            db.log(conn, "list", f"Fetch {url}")
            html = httpclient.fetch_html(session, url, referer=referer, use_cache=False)
            parsed = parser.parse_listing(html)
            added = 0
            for l in parsed["lessons"]:
                l["category_slug"] = cat
                if db.upsert_lesson_listing(conn, l):
                    added += 1
                    new_total += 1
            db.log(conn, "list",
                   f"  category={cat} page={page} found={len(parsed['lessons'])} new={added}")

            nxt = parsed["next_url"]
            if not nxt or nxt == url:
                break
            referer, url = url, nxt
            page += 1
            if page > 50:
                db.log(conn, "list", "  WARN: dừng vì >50 trang")
                break
    db.log(conn, "list", f"Tổng số bài mới thêm: {new_total}")
    return new_total


if __name__ == "__main__":
    import sys
    cats = sys.argv[1].split(",") if len(sys.argv) > 1 else None
    run(cats)
